python/ray/train/tests/test_new_persistence.py

```diff
@@ -150,6 +170,81 @@ def test_tuner(monkeypatch, storage_path_type, tmp_path):
     assert len(list(exp_dir.glob("tuner.pkl"))) == 1
 
 
+@pytest.mark.parametrize("storage_path_type", [None])
+def test_trainer(tmp_path, monkeypatch, storage_path_type):
+    """
+    TODO(justinvyu): Test for these once implemented:
+    - artifacts
+    - restoration
+    - accessing checkpoints from Result object
+    - top k checkpoints (num_to_keep)
+
+    trainer_new_persistence
+    ├── experiment_state-2023-07-28_10-00-38.json
+    ├── basic-variant-state-2023-07-28_10-00-38.json
+    ├── trainer.pkl
+    ├── tuner.pkl
+    └── DataParallelTrainer_46367_00000_0_...
+        ├── events.out.tfevents...
+        ├── params.json
+        ├── params.pkl
+        ├── progress.csv
+        ├── result.json
+        ├── checkpoint_000000
+        │   ├── dummy-0.pkl
+        │   └── dummy-1.pkl
+        ├── ...
+        ├── rank_0                  <- TODO: remove these rank folders?
+        │   ├── artifact-0-0.txt
+        │   ├── artifact-1-0.txt
+        │   └── ...
+        └── rank_1
+            ├── artifact-0-1.txt
+            ├── artifact-1-1.txt
+            └── ...
+    """
+
+    LOCAL_CACHE_DIR = tmp_path / "ray_results"
+    monkeypatch.setenv("RAY_AIR_LOCAL_CACHE_DIR", str(LOCAL_CACHE_DIR))
+    exp_name = "trainer_new_persistence"
+
+    if storage_path_type is None:
+        storage_path = None
+    elif storage_path_type == "nfs":
+        storage_path = str(tmp_path / "fake_nfs")
+
+    trainer = DataParallelTrainer(
+        train_fn,
+        train_loop_config={"in_trainer": True},
+        scaling_config=train.ScalingConfig(num_workers=2),
+        run_config=train.RunConfig(storage_path=storage_path, name=exp_name, verbose=0),
+    )
+    trainer.fit()
+
+    local_inspect_dir = tmp_path / "inspect"
+    if storage_path:
+        # if storage_filesystem:
+        #     fs, fs_path = storage_filesystem, storage_path
+        # else:
+        fs, fs_path = pyarrow.fs.FileSystem.from_uri(storage_path)
+        _download_from_fs_path(
+            fs=fs, fs_path=fs_path, local_path=str(local_inspect_dir)
+        )
+    else:
+        local_inspect_dir = LOCAL_CACHE_DIR
+
+    assert len(list(local_inspect_dir.glob("*"))) == 1  # Only expect 1 experiment dir
+    exp_dir = local_inspect_dir / exp_name
+
+    # Files synced by the driver
+    assert len(list(exp_dir.glob("basic-variant-state-*"))) == 1
+    assert len(list(exp_dir.glob("experiment_state-*"))) == 1
+    assert len(list(exp_dir.glob("tuner.pkl"))) == 1
+    assert len(list(exp_dir.glob("trainer.pkl"))) == 1
+
+    # Files synced by the worker
+
+
```

Comment:
We should check the new checkpoint's deduplication logic as well (potentially copy/adapt old test from old checkpoint).

Reply:
What's the checkpoint deduplication? Is this referring to the `to_directory` file lock?

Reply:
Yeah